x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java

```diff
@@ -181,15 +181,23 @@ public void createResponse(String docId,
                                Map<String, String> headers,
                                R response,
                                ActionListener<IndexResponse> listener) throws IOException {
-        Map<String, Object> source = new HashMap<>();
-        source.put(HEADERS_FIELD, headers);
-        source.put(EXPIRATION_TIME_FIELD, response.getExpirationTime());
-        source.put(RESULT_FIELD, encodeResponse(response));
-        IndexRequest indexRequest = new IndexRequest(index)
-            .create(true)
-            .id(docId)
-            .source(source, XContentType.JSON);
-        clientWithOrigin.index(indexRequest, listener);
+        try {
+            // TODO: Integrate with circuit breaker
```

Comment:
I wonder if `IndexRequest` could take a `ToXContent` would let you could avoid another copy.

Reply:
Great idea. I will do it in a follow up.

Reply:
I played this a bit. I am not sure if we can implement it because we need to send the [length](https://github.com/elastic/elasticsearch/blob/a32a2eaa6cb8266c54eb39851aebeb085100bbf6/server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java#L159-L160) of the source before sending its bytes. If we write ToXContent directly to the network output stream then we won't have the length in advance.